### PR TITLE
More thorough variant-api chart

### PR DESCRIPTION
**Pull request by hpatel826**
Initial draft. I will open a PR with the proper template if we decide to move forward with this.

My end-goal with these changes is to remove the need for resource templates from individual repositories. App developers only need to populate their `values.yaml`. For any runtime values, provide the values using the current pattern via terraform/octopus variables + the terraform `helm_release` resource.

### CODE REVIEW

File: temp/out.yaml [hunk 1]
```diff
@@ -0,0 +1,200 @@
+---
```
Comment (hpatel826) [inline temp/out.yaml:1]: This is the full output with the values as set in `charts/variant-api/values.yaml` and then I ran `helm template charts/variant-api/ -n schedule-adherence > temp/out.yaml` 

I also did a dry-run with the command `helm install test-release charts/variant-api --dry-run` which was successful

Please review carefully and argue 🙏

**Comment by hpatel826** [comment]
If we move forward with this, in a follow-up PR I plan to add support for the `ExternalSecret` resource to be mounted as a secret to the app container. It should be pretty seamless only requiring 1 input - the name of the AWS secret, and some well-defined place like `/app/secrets/{secret_name}` where it will be mounted as a volume. Open to other ideas that should be thrown into this chart as well.

File: temp/out.yaml [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
```
Comment (hpatel826) [inline temp/out.yaml:1]: Output generated with:

```
helm template -n schedule-adherence charts/variant-api/ --set fullnameOverride=schedule-adherence-api --set revision=abc123 --set deployment.image.tag=abc123 --set autoscaling.enabled=true --set istio.enabled=true --set serviceMonitor.enabled=true > temp/out.yaml 
```

File: charts/variant-api/templates/_helpers.tpl [hunk 3]
```diff
@@ -40,6 +40,7 @@ helm.sh/chart: {{ include "chart.chart" . }}
 app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
```
[the remark this answers is not in the document]
Comment (anr-variant) [reply charts/variant-api/templates/_helpers.tpl:40]: This seems good with me since it's just the regular labels and not the selectors that we'd be adding this to.

For reference, Tesh and I had previously contemplated adding revision to selector labels, but we determined that it was a bigger risk than necessary because the service wouldn't be able to point to existing pods while the deployment was rolling out or wouldn't be able to point to new pods while the rollout was happening, depending on when Helm decided to modify the service. There was the situation where a deployment rollout could fail because the pod refused to start for some reason and then the service only has a failed pod to point at.

File: charts/variant-api/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      {{- $selectorLabels | nindent 6 }}
+  template:
+    metadata:
+      labels:
+        {{- $selectorLabels | nindent 8 }}
+    spec:
+      serviceAccountName: {{ $fullName }}
+      automountServiceAccountToken: true
+      securityContext:
+        fsGroup: 65534
```
Comment (anr-variant) [inline charts/variant-api/templates/deployment.yaml:22]: Probably worth adding the clarifying comment that this the `nobody` group for alpine. If we're standardized on alpine everywhere hardcoding this fine, if not we probably want this to be a variable because there's actually no standard for it 🙃 It's 99 on some linux distros though I'm not sure which ones. Went digging to see what it is debian slim and ubuntu (since those seem like the next most likely candidates, especially if we have any Python containers being deployed since `muslc` is hell for Python last I looked into it) but not really finding anything concrete.

File: charts/variant-api/templates/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+          readinessProbe:
+            httpGet:
+              path: /health
+              port: http
+            initialDelaySeconds: 10
+            periodSeconds: 10
+          resources:
+            {{- toYaml .Values.deployment.resources | nindent 12 }}
+          {{- if len .Values.deployment.envVars }}
+          env:
```
Comment (anr-variant) [inline charts/variant-api/templates/deployment.yaml:52]: Probably worth throwing `revision` into here as well since we need to expose it via health checks and that makes it easy to grab from a known, standard environment variable.

File: charts/variant-api/templates/service-monitor.yaml [hunk 6]
```diff
@@ -7,14 +8,16 @@ metadata:
   name: {{ include "chart.fullname" . }}
 spec:
   endpoints:
-    - targetPort: {{ default 9090 .Values.serviceMonitor.targetPort }}
-      port: {{ default "metrics" .Values.serviceMonitor.port }}
+    - targetPort: {{ default 9090 .Values.service.targetPort }}
+      port: {{ default "metrics" .Values.service.port }}
```
Comment (anr-variant) [inline charts/variant-api/templates/service-monitor.yaml:12]: From the documentation, `port` and `targetPort` are mutually exclusive and `port` seems to be the preferred since it references a named port.

https://github.com/prometheus-operator/prometheus-operator/blob/c6b5cc2e178002683ab3afe8d023a8bb2508ba19/Documentation/api.md#endpoint

File: charts/variant-api/templates/service.yaml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+{{- $labels := (include "chart.labels" .) -}}
+{{- $selectorLabels := (include "chart.selectorLabels" .) -}}
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ $fullName }}
+  labels:
+    {{- $labels | nindent 4 }}
+spec:
+  type: {{ .Values.service.type }}
```
Comment (anr-variant) [inline charts/variant-api/templates/service.yaml:11]: Part of me wants to say we just hardcode this `ClusterIp` to prevent someone accidentally creating a `LoadBalancer` or `NodePort`. Not sure though. However, I can't think of a situation for our APIs that would need anything other than `ClusterIp` since the only other option is `ExternalName` which is just used for creating what are effectively CNAME records.

File: charts/variant-api/templates/serviceaccount.yaml [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+{{- $labels := (include "chart.labels" .) -}}
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ $fullName }}
+  {{- if .Values.serviceAccount.roleArn }}
+  labels:
+    {{- $labels | nindent 4 }}
+  annotations:
+    "eks.amazonaws.com/role-arn": "{{ .Values.serviceAccount.roleArn }}"
```
Comment (anr-variant) [inline charts/variant-api/templates/serviceaccount.yaml:11]: Should we allow other annotations onto this? We don't use them currently but have a feeling as soon as this is merged and sees usage someone will come up with a use case lol.

**Review by anr-variant — COMMENTED** [review]
Seems like the big idea behind here is that we can all delete our helm directories, update the `helm_release` terraform block to point at where this chart lives and then just machine gun `set` blocks at it to configure this how we need it, is that accurate?

**Comment by hpatel826** [comment]
> Seems like the big idea behind here is that we can all delete our helm directories (for APIs at least), update the `helm_release` terraform block to point at where this chart lives and then just machine gun `set` blocks at it to configure this how we need it, is that accurate?

Yes. Want to follow up with the same idea for `variant-handler` and eventually for cronjob/job as needed.

File: charts/variant-api/templates/service.yaml (hunk 7 above)
Comment (hpatel826) [reply charts/variant-api/templates/service.yaml:11]: I'm trying to minimize the thought needed by devs when using these charts for people who don't want to look at K8S object definitions. Based on your comment I think we should hard code as `ClusterIP`, don't expose the setting at all to be pondered by a dev.

File: charts/variant-api/templates/deployment.yaml (hunk 4 above)
Comment (hpatel826) [reply charts/variant-api/templates/deployment.yaml:22]: IMO people using these "all in one" charts should just be on alpine so this doesn't have to be considered when using the chart, but I can't think of a good way to enforce/ensure that from the chart since there is no real context for the image ...

File: temp/out.yaml (hunk 2 above)
Comment (hpatel826) [reply temp/out.yaml:1]: More SETs:

```
helm template -n schedule-adherence charts/variant-api/ --set fullnameOverride=schedule-adherence-api --set revision=abc123 --set deployment.image.tag=abc123 --set autoscaling.enabled=true --set istio.enabled=true --set serviceMonitor.enabled=true --set "istio.egress[0].name=test" --set "istio.egress[0].hosts={test.com,blah.net}" --set "istio.egress[0].ports[0].number=80" --set "istio.egress[0].ports[0].protocol=HTTP" > temp/out.yaml
 ```

File: charts/variant-api/templates/deployment.yaml (hunk 4 above)
Comment (anr-variant) [reply charts/variant-api/templates/deployment.yaml:22]: Yeah. I'm fine leaving it as is and if someone comes along with like a `debian-slim` application that this fails on then we address it then. 

I do think leaving a comment stating it's the nobody group is a good idea though.

File: charts/variant-api/templates/serviceaccount.yaml (hunk 8 above)
Comment (anr-variant) [reply charts/variant-api/templates/serviceaccount.yaml:11]: Actually on second though, pod annotations are probably the only real use for annotations right now. At least I had a use case for them to test the istio `excludeOutboundPorts` stuff, but that's also kind of a weird thing I was doing too. If I had to pick just one I think I'd pick pod annotations since the role-arn is the only one we've ever applied from an application to a service account.

File: charts/variant-api/templates/deployment.yaml [hunk 9]
```diff
@@ -22,6 +22,24 @@ spec:
...
+      # fsGroup determines group ownership of volumes mounted in this dynamic manner.
+      #
+      # Assumptions:
+      # 1) The application process runs on an Alpine based platform -- the "nobody" group is defined as ID 65534 in Alpine
+      # 2) The application is running as a non-root user (u better be)
+      # 3) The non-root user _does not_ belong to any groups (i.e, "nobody")
+      #
+      # Let's not run as root
+      # Chown all the things that you own
+      # Groups of nobody
```
Comment (anr-variant) [inline charts/variant-api/templates/deployment.yaml:41]: <3

**Comment by cjackson-variant** [comment]
LGTM. I like having this stuff in 1 place.

File: temp/out.yaml [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
```
Comment (hpatel826) [inline temp/out.yaml:1]: Output generated with this command: 

`helm template -f temp/test-values.yaml -n schedule-adherence charts/variant-api/ > temp/out.yaml`

File: charts/variant-api/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+          readinessProbe:
+            httpGet:
+              path: /health
+              port: http
+            initialDelaySeconds: 10
+            periodSeconds: 10
+          resources:
+            {{- toYaml .Values.deployment.resources | nindent 12 }}
+          env:
+            - name: REVISION
```
Comment (vibindaniel) [inline charts/variant-api/templates/deployment.yaml:74]: Any reason why revision is required?

File: charts/variant-api/templates/_helpers.tpl [hunk 12]
```diff
@@ -1,26 +1,10 @@
...
-{{- .Values.fullnameOverride | trunc 63 | trimSuffix "-" }}
-{{- else }}
-{{- $name := default .Chart.Name .Values.nameOverride }}
-{{- if contains $name .Release.Name }}
-{{- .Release.Name | trunc 63 | trimSuffix "-" }}
-{{- else }}
-{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" }}
-{{- end }}
-{{- end }}
+{{- required "fullnameOverride is required" .Values.fullnameOverride | trunc 63 | trimSuffix "-" }}
```
Comment (vibindaniel) [inline charts/variant-api/templates/_helpers.tpl:7]: Why is this required?

Comment (hpatel826) [reply charts/variant-api/templates/_helpers.tpl:7]: @vibindaniel this is meant to be the 1 single name input for the API project being deployed. This is the base name used for all resources created by the chart. I could probably just rename this value as "apiName" or just "name". Thoughts? 

The goal is to have just 1 required name input used everywhere and to limit the amount of logic/lookup when future developers look at this chart to understand how things are deployed and named.

File: charts/variant-api/templates/deployment.yaml (hunk 11 above)
Comment (hpatel826) [reply charts/variant-api/templates/deployment.yaml:74]: For people outside of the owning team. A quick answer to the question "what  created or modified this resource most recently". The question is answered the same way for all chart deployments if we require it. 

I can make it optional. Maybe @anr-variant has some thoughts?

File: charts/variant-api/templates/_helpers.tpl (hunk 12 above)
Comment (hpatel826) [reply charts/variant-api/templates/_helpers.tpl:7]: Makes sense to me!! I'll replace that here and update the docs. 1 less input!!

File: charts/variant-api/templates/deployment.yaml (hunk 11 above)
Comment (vibindaniel) [reply charts/variant-api/templates/deployment.yaml:74]: Can't that be part of the label instead?

Comment (hpatel826) [reply charts/variant-api/templates/deployment.yaml:74]: It is part of the label too in `_helpers.tpl`.  @anr-variant requested it as an `env` so that applications can simply pull in the environment variable instead of being K8S-aware

File: charts/variant-api/HELM_CLI_USAGE.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
+# Helm CLI Usage
+
+To install the chart,
+first add the repo
+
+`helm repo add variant-inc-helm-charts https://variant-inc.github.io/lazy-helm-charts/ --password <token>`
+
+and then install the chart using
+
+`helm upgrade --install my-api variant-inc-helm-charts/variant-api -n my-namespace --set "istio.ingress.host=dev-drivevariant.com" --set "revision=abc123" --set "deployment.image.tag=ecr.amazonaws.com/my-project/my-api:abc123"`
```
Comment (cjackson-variant) [inline charts/variant-api/HELM_CLI_USAGE.md:10]: Does this exact command have to be run or is this an example? I can set any relevant options defined in the readme right?

File: charts/variant-api/MIGRATING.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+  values = [<<EOF
+{chart alias}:
+  deployment:
+    envVars:
+      - name: CONNECTIONSTRINGS__REDIS
+        value: "${var.redis_host}:${var.redis_port}"
+      - name: HERECLIENT__CACHE_DURATION_MINUTES
+        value: var.here_cache_duration_minutes
+      - name: EPSAGON_TOKEN
+        value: ${var.epsagon_token}
```
Comment (cjackson-variant) [inline charts/variant-api/MIGRATING.md:91]: It would be good to add notes about the variable interpolation syntax here.

File: charts/variant-api/README.md [hunk 15]
```diff
@@ -1,27 +1,109 @@
...
+  - See [infrastructure permissions](#infrastructure-permissions)
+
+## Before you start
+
+1. Use a CloudOps Github CI workflow that publishes an image
+   - [.NET](https://github.com/variant-inc/actions-dotnet)
+   - [Node](https://github.com/variant-inc/actions-nodejs)
+   - [Python](https://github.com/variant-inc/actions-python)
+1. Host a health check endpoint via `GET /health` which returns a status code < 400 when healthy or >= 400 when unhealthy
+1. Host a Prometheus metrics endpoint via `GET /metrics`
```
Comment (cjackson-variant) [inline charts/variant-api/README.md:62]: It looks like we're moving to Epsagon for metrics correct?

**Comment by hpatel826** [comment]
Releasing this chart as beta pending some potential changes around metrics + optional cron curl jobs to support some other non-engineering team workflow
